Pytest aborts when fixture errors during teardown and `--maxfail=1`

- [x] a detailed description of the bug or problem you are having
- [x] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

## Description

Pytest aborts if a session-scoped fixture errors during teardown when (a) one or more test cases fail and (b) `--maxfail=1` is used.

When this happens, `test-report.xml` does not contain the teardown fixture error. This is problematic as users may want to also record teardown fixture errors when their test cases fail.

## Minimal Example

### Code

```python
import pytest


def some_condition() -> bool:
    return True


@pytest.fixture(scope="session", autouse=True)
def my_end_of_session_check():
    yield

    print("Running end of session check!")
    if some_condition():
        pytest.fail("Oh noes")


def test_case_1():
    pytest.fail("This is a failing test")


def test_case_2():
     pytest.fail("This is also a failing test")
```

### Args

`pytest example.py --maxfail=1 --junitxml=test-report.xml`

### Output

Console:

```txt
============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.8.13, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/bbrown/Downloads
collected 2 items

example.py FRunning end of session check!
Traceback (most recent call last):
  File "/Users/bbrown/Downloads/.venv/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  <clipped>
  File "/Users/bbrown/Downloads/.venv/lib/python3.8/site-packages/_pytest/fixtures.py", line 911, in _teardown_yield_fixture
    next(it)
  File "/Users/bbrown/Downloads/example.py", line 14, in my_end_of_session_check
    pytest.fail("Oh noes")
  File "/Users/bbrown/Downloads/.venv/lib/python3.8/site-packages/_pytest/outcomes.py", line 198, in fail
    raise Failed(msg=reason, pytrace=pytrace)
Failed: Oh noes
```

`test-report.xml`:

```xml
<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite name="pytest" errors="0" failures="1" skipped="0" tests="1" time="0.035" timestamp="2023-12-15T08:56:19.980572" hostname="macbookpro"><testcase classname="example" name="test_case_1" time="0.001"><failure message="Failed: This is a failing test">def test_case_1():
&gt;       pytest.fail("This is a failing test")
E       Failed: This is a failing test

example.py:18: Failed</failure></testcase></testsuite></testsuites>
```

### Expected Output

I expect pytest to not abort, and report "1 failed, 1 error, 1 skipped" in both the console and `test-report.xml` when this happens.

### Environment

__OS:__

macOS Sonoma 14.1.2

__Virtual environment:__

```txt
(.venv) % pip list
Package        Version
-------------- -------
exceptiongroup 1.2.0
iniconfig      2.0.0
packaging      23.2
pip            23.3.1
pluggy         1.3.0
pytest         7.4.3
setuptools     56.0.0
tomli          2.0.1
```

--- a/AUTHORS
+++ b/AUTHORS
@@ -54,6 +54,7 @@ Aviral Verma
 Aviv Palivoda
 Babak Keyvani
 Barney Gale
+Ben Brown
 Ben Gartner
 Ben Webb
 Benjamin Peterson

--- a/changelog/11706.bugfix.rst
+++ b/changelog/11706.bugfix.rst
@@ -1,0 +1,1 @@
+Fix reporting of teardown errors in session and module scoped fixtures when `--maxfail=1`.

--- a/changelog/11706.bugfix.rst
+++ b/changelog/11706.bugfix.rst
@@ -1,1 +1,1 @@
-Fix reporting of teardown errors in session and module scoped fixtures when `--maxfail=1`.
+Fix reporting of teardown errors in higher-scoped fixtures when using `--maxfail` or `--stepwise`.

--- a/src/_pytest/main.py
+++ b/src/_pytest/main.py
@@ -548,8 +548,8 @@ class Session(nodes.Collector):
         )
         self.testsfailed = 0
         self.testscollected = 0
-        self.shouldstop: Union[bool, str] = False
-        self.shouldfail: Union[bool, str] = False
+        self._shouldstop: Union[bool, str] = False
+        self._shouldfail: Union[bool, str] = False
         self.trace = config.trace.root.get("collection")
         self._initialpaths: FrozenSet[Path] = frozenset()
         self._initialpaths_with_parents: FrozenSet[Path] = frozenset()
@@ -575,6 +575,28 @@ class Session(nodes.Collector):
             self.testsfailed,
             self.testscollected,
         )
+
+    @property
+    def shouldstop(self) -> Union[bool, str]:
+        return self._shouldstop
+
+    @shouldstop.setter
+    def shouldstop(self, value: Union[bool, str]) -> None:
+        """Prevent plugins from setting this to False once it has been set."""
+        if value is False and self._shouldstop:
+            return
+        self._shouldstop = value
+
+    @property
+    def shouldfail(self) -> Union[bool, str]:
+        return self._shouldfail
+
+    @shouldfail.setter
+    def shouldfail(self, value: Union[bool, str]) -> None:
+        """Prevent plugins from setting this to False once it has been set."""
+        if value is False and self._shouldfail:
+            return
+        self._shouldfail = value
 
     @property
     def startpath(self) -> Path:

--- a/src/_pytest/main.py
+++ b/src/_pytest/main.py
@@ -584,7 +584,8 @@ class Session(nodes.Collector):
 
     @shouldstop.setter
     def shouldstop(self, value: Union[bool, str]) -> None:
-        """Prevent plugins from setting this to False once it has been set."""
+        # The runner checks shouldfail and assumes that if it is set we are
+        # definitely stopping, so prevent unsetting it.
         if value is False and self._shouldstop:
             warnings.warn(
                 PytestWarning(
@@ -601,7 +602,8 @@ class Session(nodes.Collector):
 
     @shouldfail.setter
     def shouldfail(self, value: Union[bool, str]) -> None:
-        """Prevent plugins from setting this to False once it has been set."""
+        # The runner checks shouldfail and assumes that if it is set we are
+        # definitely stopping, so prevent unsetting it.
         if value is False and self._shouldfail:
             warnings.warn(
                 PytestWarning(

--- a/src/_pytest/main.py
+++ b/src/_pytest/main.py
@@ -6,6 +6,7 @@ import functools
 import importlib
 import os
 import sys
+import warnings
 from pathlib import Path
 from typing import AbstractSet
 from typing import Callable
@@ -44,6 +45,7 @@ from _pytest.reports import CollectReport
 from _pytest.reports import TestReport
 from _pytest.runner import collect_one_node
 from _pytest.runner import SetupState
+from _pytest.warning_types import PytestWarning
 
 
 def pytest_addoption(parser: Parser) -> None:
@@ -584,6 +586,12 @@ class Session(nodes.Collector):
     def shouldstop(self, value: Union[bool, str]) -> None:
         """Prevent plugins from setting this to False once it has been set."""
         if value is False and self._shouldstop:
+            warnings.warn(
+                PytestWarning(
+                    "session.shouldstop cannot be unset after it has been set; ignoring."
+                ),
+                stacklevel=2,
+            )
             return
         self._shouldstop = value
 
@@ -595,6 +603,12 @@ class Session(nodes.Collector):
     def shouldfail(self, value: Union[bool, str]) -> None:
         """Prevent plugins from setting this to False once it has been set."""
         if value is False and self._shouldfail:
+            warnings.warn(
+                PytestWarning(
+                    "session.shouldfail cannot be unset after it has been set; ignoring."
+                ),
+                stacklevel=2,
+            )
             return
         self._shouldfail = value
 

--- a/src/_pytest/runner.py
+++ b/src/_pytest/runner.py
@@ -131,6 +131,12 @@ def runtestprotocol(
             show_test_item(item)
         if not item.config.getoption("setuponly", False):
             reports.append(call_and_report(item, "call", log))
+
+    # If the session is about to fail or stop, teardown everything - this is
+    # necessary to correctly report fixture teardown errors (see #11706)
+    if item.session.shouldfail or item.session.shouldstop:
+        nextitem = None
+
     reports.append(call_and_report(item, "teardown", log, nextitem=nextitem))
     # After all teardown hooks have been called
     # want funcargs and request info to go away.
@@ -179,9 +185,6 @@ def pytest_runtest_call(item: Item) -> None:
 
 def pytest_runtest_teardown(item: Item, nextitem: Optional[Item]) -> None:
     _update_current_test_var(item, "teardown")
-    # If the session is about to fail, teardown everything - this is necessary
-    # to correctly report fixture teardown errors (see #11706)
-    nextitem = None if item.session.shouldfail else nextitem
     item.session._setupstate.teardown_exact(nextitem)
     _update_current_test_var(item, None)
 

--- a/src/_pytest/runner.py
+++ b/src/_pytest/runner.py
@@ -131,12 +131,10 @@ def runtestprotocol(
             show_test_item(item)
         if not item.config.getoption("setuponly", False):
             reports.append(call_and_report(item, "call", log))
-
     # If the session is about to fail or stop, teardown everything - this is
     # necessary to correctly report fixture teardown errors (see #11706)
     if item.session.shouldfail or item.session.shouldstop:
         nextitem = None
-
     reports.append(call_and_report(item, "teardown", log, nextitem=nextitem))
     # After all teardown hooks have been called
     # want funcargs and request info to go away.

--- a/src/_pytest/runner.py
+++ b/src/_pytest/runner.py
@@ -179,6 +179,9 @@ def pytest_runtest_call(item: Item) -> None:
 
 def pytest_runtest_teardown(item: Item, nextitem: Optional[Item]) -> None:
     _update_current_test_var(item, "teardown")
+    # If the session is about to fail, teardown everything - this is necessary
+    # to correctly report fixture teardown errors (see #11706)
+    nextitem = None if item.session.shouldfail else nextitem
     item.session._setupstate.teardown_exact(nextitem)
     _update_current_test_var(item, None)
 

--- a/testing/test_runner.py
+++ b/testing/test_runner.py
@@ -1087,3 +1087,25 @@ def test_outcome_exception_bad_msg() -> None:
     with pytest.raises(TypeError) as excinfo:
         OutcomeException(func)  # type: ignore
     assert str(excinfo.value) == expected
+
+
+def test_teardown_session_failed(pytester: Pytester) -> None:
+    """Test that fixture teardown failures are reported after a test fails."""
+    pytester.makepyfile(
+        """
+        import pytest
+
+        @pytest.fixture(scope="module")
+        def baz():
+            yield
+            pytest.fail("This is a failing fixture")
+
+        def test_foo(baz):
+            pytest.fail("This is a failing test")
+
+        def test_bar(baz):
+            pass
+    """
+    )
+    result = pytester.runpytest("--maxfail=1")
+    result.stdout.fnmatch_lines(["*1 failed, 1 error*"])

--- a/testing/test_runner.py
+++ b/testing/test_runner.py
@@ -1090,7 +1090,11 @@ def test_outcome_exception_bad_msg() -> None:
 
 
 def test_teardown_session_failed(pytester: Pytester) -> None:
-    """Test that fixture teardown failures are reported after a test fails."""
+    """Test that higher-scoped fixture teardowns run in the context of the last
+    item after the test session bails early due to --maxfail.
+
+    Regression test for #11706.
+    """
     pytester.makepyfile(
         """
         import pytest
@@ -1108,4 +1112,4 @@ def test_teardown_session_failed(pytester: Pytester) -> None:
     """
     )
     result = pytester.runpytest("--maxfail=1")
-    result.stdout.fnmatch_lines(["*1 failed, 1 error*"])
+    result.assert_outcomes(failed=1, errors=1)

--- a/testing/test_runner.py
+++ b/testing/test_runner.py
@@ -1102,22 +1102,21 @@ def test_teardown_session_failed(pytester: Pytester) -> None:
         @pytest.fixture(scope="module")
         def baz():
             yield
-            pytest.fail("This is a failing fixture")
+            pytest.fail("This is a failing teardown")
 
         def test_foo(baz):
             pytest.fail("This is a failing test")
 
-        def test_bar(baz):
-            pass
-    """
+        def test_bar(): pass
+        """
     )
     result = pytester.runpytest("--maxfail=1")
     result.assert_outcomes(failed=1, errors=1)
 
 
 def test_teardown_session_stopped(pytester: Pytester) -> None:
-    """Similar to `test_teardown_session_failed`, but uses `pytest.exit` instead
-    of `pytest.fail`.
+    """Test that higher-scoped fixture teardowns run in the context of the last
+    item after the test session bails early due to --stepwise.
 
     Regression test for #11706.
     """
@@ -1128,14 +1127,13 @@ def test_teardown_session_stopped(pytester: Pytester) -> None:
         @pytest.fixture(scope="module")
         def baz():
             yield
-            pytest.exit("This is an exiting fixture")
+            pytest.fail("This is a failing teardown")
 
         def test_foo(baz):
             pytest.fail("This is a failing test")
 
-        def test_bar(baz):
-            pass
-    """
+        def test_bar(): pass
+        """
     )
-    result = pytester.runpytest("--maxfail=1")
-    result.assert_outcomes(failed=1, errors=0)
+    result = pytester.runpytest("--stepwise")
+    result.assert_outcomes(failed=1, errors=1)

--- a/testing/test_runner.py
+++ b/testing/test_runner.py
@@ -1113,3 +1113,29 @@ def test_teardown_session_failed(pytester: Pytester) -> None:
     )
     result = pytester.runpytest("--maxfail=1")
     result.assert_outcomes(failed=1, errors=1)
+
+
+def test_teardown_session_stopped(pytester: Pytester) -> None:
+    """Similar to `test_teardown_session_failed`, but uses `pytest.exit` instead
+    of `pytest.fail`.
+
+    Regression test for #11706.
+    """
+    pytester.makepyfile(
+        """
+        import pytest
+
+        @pytest.fixture(scope="module")
+        def baz():
+            yield
+            pytest.exit("This is an exiting fixture")
+
+        def test_foo(baz):
+            pytest.fail("This is a failing test")
+
+        def test_bar(baz):
+            pass
+    """
+    )
+    result = pytester.runpytest("--maxfail=1")
+    result.assert_outcomes(failed=1, errors=0)

--- a/testing/test_session.py
+++ b/testing/test_session.py
@@ -418,3 +418,27 @@ def test_rootdir_wrong_option_arg(pytester: Pytester) -> None:
     result.stderr.fnmatch_lines(
         ["*Directory *wrong_dir* not found. Check your '--rootdir' option.*"]
     )
+
+
+def test_shouldfail_is_sticky(pytester: Pytester) -> None:
+    """Test that session.shouldfail cannot be reset to False after being set."""
+    pytester.makeconftest(
+        """
+        def pytest_sessionfinish(session, exitstatus):
+            session.shouldfail = False
+    """
+    )
+
+    pytester.makepyfile(
+        """
+        import pytest
+
+        def test_foo():
+            pytest.fail("This is a failing test")
+
+        def test_bar():
+            pass
+    """
+    )
+    result = pytester.runpytest("--maxfail=1")
+    result.stderr.fnmatch_lines("*session.shouldfail cannot be unset*")

--- a/testing/test_session.py
+++ b/testing/test_session.py
@@ -421,14 +421,18 @@ def test_rootdir_wrong_option_arg(pytester: Pytester) -> None:
 
 
 def test_shouldfail_is_sticky(pytester: Pytester) -> None:
-    """Test that session.shouldfail cannot be reset to False after being set."""
+    """Test that session.shouldfail cannot be reset to False after being set.
+
+    Issue #11706.
+    """
     pytester.makeconftest(
         """
-        def pytest_sessionfinish(session, exitstatus):
+        def pytest_sessionfinish(session):
+            assert session.shouldfail
             session.shouldfail = False
-    """
+            assert session.shouldfail
+        """
     )
-
     pytester.makepyfile(
         """
         import pytest
@@ -436,9 +440,41 @@ def test_shouldfail_is_sticky(pytester: Pytester) -> None:
         def test_foo():
             pytest.fail("This is a failing test")
 
-        def test_bar():
-            pass
-    """
+        def test_bar(): pass
+        """
     )
-    result = pytester.runpytest("--maxfail=1")
-    result.stderr.fnmatch_lines("*session.shouldfail cannot be unset*")
+
+    result = pytester.runpytest("--maxfail=1", "-Wall")
+
+    result.assert_outcomes(failed=1, warnings=1)
+    result.stdout.fnmatch_lines("*session.shouldfail cannot be unset*")
+
+
+def test_shouldstop_is_sticky(pytester: Pytester) -> None:
+    """Test that session.shouldstop cannot be reset to False after being set.
+
+    Issue #11706.
+    """
+    pytester.makeconftest(
+        """
+        def pytest_sessionfinish(session):
+            assert session.shouldstop
+            session.shouldstop = False
+            assert session.shouldstop
+        """
+    )
+    pytester.makepyfile(
+        """
+        import pytest
+
+        def test_foo():
+            pytest.fail("This is a failing test")
+
+        def test_bar(): pass
+        """
+    )
+
+    result = pytester.runpytest("--stepwise", "-Wall")
+
+    result.assert_outcomes(failed=1, warnings=1)
+    result.stdout.fnmatch_lines("*session.shouldstop cannot be unset*")

